dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py

```diff
@@ -0,0 +1,207 @@
+        self.mode = mode
+        self.num_cached_batches = num_cached_batches
+        self.metrics_file = metrics_file
+
+        # In-memory cache for batches
+        self.cached_batches = deque(maxlen=num_cached_batches)
+        self.cache_ready = False
+        self.cache_lock = threading.Lock()
+
+        # Performance metrics
+        self.batch_times = []
+        self.start_time = None
+        self.end_time = None
+
+        # If replay mode, cache batches during construction
+        if self.mode == "replay":
+            self._cache_batches()
+
+    def __iter__(self) -> Iterator[Any]:
+        """Iterate through the dataloader based on the current mode."""
+        if self.mode == "log":
+            return self._log_mode_iter()
+        elif self.mode == "replay":
+            return self._replay_mode_iter()
+        else:
+            # Default to log mode for unknown modes
+            return self._log_mode_iter()
+
+    def __len__(self):
+        """Return the length of the wrapped dataloader."""
+        return len(self.dataloader)
+
+    def _cache_batches(self):
+        """Cache batches from the dataloader for replay mode."""
+        with self.cache_lock:
+            self.cached_batches.clear()
+            self.cache_ready = False
+
+        for batch in self.dataloader:
+            with self.cache_lock:
+                self.cached_batches.append(batch)
+                # Cache is ready when we have at least one batch cached
+                if len(self.cached_batches) > 0:
+                    self.cache_ready = True
+
+    def _log_mode_iter(self) -> Iterator[Any]:
+        """Log mode: iterate normally while collecting metrics."""
+        self.start_time = time.time()
+        self.batch_times = []
+
+        dataloader_iter = iter(self.dataloader)
+        while True:
+            try:
+                batch_start = time.time()
+                batch = next(dataloader_iter)  # This is where the actual data loading happens
+                batch_time = time.time() - batch_start
+                self.batch_times.append(batch_time)
+                yield batch
+            except StopIteration:
+                break
+
+        self.end_time = time.time()
+
+    def _replay_mode_iter(self) -> Iterator[Any]:
+        """Replay mode: replay cached batches for ideal performance
+        simulation."""
+        if not self.cache_ready or len(self.cached_batches) == 0:
+            raise RuntimeError(
+                "No cached batches available. This should not happen in\n"
+                "replay mode as batches are cached during construction."
+            )
+
+        self.start_time = time.time()
+        self.batch_times = []
+
+        # Replay cached batches to match original DataLoader length
+        original_length = len(self.dataloader)
+
+        for i in range(original_length):
+            batch_start = time.time()
+            batch = self.cached_batches[i % len(self.cached_batches)]
```

Comment:
shouldn't you use `self.cache_lock` here as well, or is it guaranteed that only one thread can access the cached_batches structure once in replay mode?